flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java

```diff
@@ -140,6 +141,12 @@ public UpdateControl<FlinkDeployment> reconcile(FlinkDeployment flinkApp, Contex
                         previousDeployment,
                         false);
             }
+            // Rely on the last stable spec if rolling back
+            if (flinkApp.getStatus().getReconciliationStatus().getState()
+                    == ReconciliationState.ROLLING_BACK) {
+                flinkApp.setSpec(
+                        flinkApp.getStatus().getReconciliationStatus().deserializeLastStableSpec());
```

Comment:
We might want to do a spec diff check there with the current spec vs the last reconciled spec. It could happen that we initiated the rollback (setting the state to ROLLING_BACK) and in the meantime the user actually submitted a new spec.

With the current logic the new spec would be first ignored and we perform a rollback and after that we reconcile the new spec. This is not entirely incorrect but it will lead to extra time and extra redeployments.

Reply:
I've updated the PR to take in account this case.

I also set the state to Upgrading if the there is a spec change as we don't want to apply specific code path related to roll back mechanism: https://github.com/apache/flink-kubernetes-operator/pull/622/files#diff-29ea38a50cac5b4432dd0969bc3e2177e29a5507f8c7bb01b80f605a8740de41R146

I had here to somehow exclude a specific spec diff when it is due to job.state as we can override this job.state in https://github.com/apache/flink-kubernetes-operator/blob/main/flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java#L358

Reply:
My change is bad for now there are still some cases with job.state diff and some usually ignore changed (for ex on `kubernetes.operator.deployment.rollback.enabled`) that are seen as an upgrade change now and lead to an upgrade instead of roll abck

Reply:
I will take a look, it could be related to the latest changes we have made to the master regarding the rescaling logic.

Reply:
For this issue it is clearly on my side.
I should have found a good way to detect the case I'm currently rebasing from last master (few conflict to resolve) and will push it

Reply:
Rebased + updated with [working solution](https://github.com/apache/flink-kubernetes-operator/pull/622/files#diff-29ea38a50cac5b4432dd0969bc3e2177e29a5507f8c7bb01b80f605a8740de41R157)

Reply:
The getAvailableUpgrade mode method would only return save point if the job is either already shut down / terminally failed or is running and can be upgraded with the save point. So if it returns SAVEPOINT it should be safe to call cancel with savepoint.